File: pkgs/core/tigrbl_typing/tigrbl_typing/channel.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, Literal, Mapping, Sequence

OpChannelKind = Literal["http", "websocket", "sse", "stream", "webtransport"]
OpChannelFamily = Literal["request", "response", "stream", "socket", "session"]
OpChannelSubevent = Literal[
    "connect",
    "receive",
    "emit",
    "complete",
    "disconnect",
]


@dataclass(slots=True)
class OpChannel:
    """Runtime transport descriptor shared across Python and Rust surfaces."""

    kind: OpChannelKind
    family: OpChannelFamily
    exchange: str
    protocol: str
    path: str
    method: str | None = None
    selector: str | None = None
    framing: str | None = None
    subevents: tuple[OpChannelSubevent, ...] = ()
    headers: Mapping[str, str] = field(default_factory=dict)
    query: Mapping[str, Sequence[str]] = field(default_factory=dict)
    path_params: Mapping[str, str] = field(default_factory=dict)
    state: dict[str, Any] = field(default_factory=dict)
    send: Callable[[Any], Any] | None = None
    receive: Callable[[], Any] | None = None


__all__ = [
    "OpChannel",
    "OpChannelFamily",
    "OpChannelKind",
    "OpChannelSubevent",
]

File: crates/tigrbl_rs_ports/src/channel.rs
use crate::errors::PortResult;
use tigrbl_rs_spec::values::Value;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ChannelFamily {
    Request,
    Response,
    Stream,
    Socket,
    Session,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ChannelSubevent {
    Connect,
    Receive,
    Emit,
    Complete,
    Disconnect,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct OpChannel {
    pub kind: String,
    pub family: ChannelFamily,
    pub exchange: String,
    pub protocol: String,
    pub path: String,
    pub selector: Option<String>,
    pub framing: Option<String>,
    pub subevents: Vec<ChannelSubevent>,
}

pub trait ChannelPort: Send + Sync {
    fn receive(&self, channel: &OpChannel) -> PortResult<Option<Value>>;
    fn emit(&self, channel: &OpChannel, payload: Value) -> PortResult<()>;
    fn complete(&self, channel: &OpChannel) -> PortResult<()>;
}
